assumeutxo: Add dumptxoutset height param, remove shell scripts

fjahr opened the pull request:
This adds a height parameter to the `dumptxoutset` RPC. This internalizes the workflow that was previously done by scripts: roll back the chain to the height we actually want the snapshot from, create the snapshot, roll forward to the real tip again.

The nice thing about internalizing this functionality is that we can write tests for the code and it gives us more options to make the functionality robust. The shell scripts we have so far will be more cumbersome to maintain in the long run, especially since we will only notice later when we have broken them. I think it's safe to remove these `test_utxo_snapshots.sh` as well when we have this option in `dumptxoutset` because we have also added some good additional functional test coverage for this functionality.

fjahr committed:
RPC: Extract InvalidateBlock helper

fjahr committed:
RPC: Extract ReconsiderBlock helper

fjahr committed:
RPC: Add type parameter to dumptxoutset

fjahr committed:
test: Test for dumptxoutset at specific height

fjahr committed:
contrib: Remove test_utxo_snapshots.sh

fjahr committed:
assumeutxo: Remove devtools/utxo_snapshot.sh

Sjors commented:
Concept ACK. But rather than relying on block invalidation, I would prefer to have a clean rollback. That would be useful as it's own RPC too: turn off network, call `rollback N/HASH`, turn on network and it will sync normally. See #29565.

TheCharlatan commented:
Concept ACK

BrandonOdiwuor reviewed (commented):
Concept ACK

great idea moving `InvalidateBlock` and `ReconsiderBlock` to the `dumputxoset` RPC instead of relying on shell scripts

fjahr committed:
doc: Improve assumeutxo guide and add more docs/comments

Also fixes some outdated information in the remaining design doc.

fjahr commented on src/rpc/blockchain.cpp:
```diff
@@ -2621,10 +2624,30 @@ static RPCHelpMan dumptxoutset()
+        const CBlockIndex* new_tip_index{WITH_LOCK(::cs_main, return node.chainman->ActiveChain().Tip())};
+        // In case there is any issue with a block being read from disc due to
+        // pruning for example, we need to stop here, otherwise the dump would
+        // still be greated for the wrong height.
```
fixed

fjahr commented on test/functional/feature_assumeutxo.py:
```diff
@@ -210,6 +213,13 @@ def run_test(self):
 
         assert_equal(n0.getblockchaininfo()["blocks"], FINAL_HEIGHT)
 
+        self.log.info(f"Check that dumptxoutsetinfo works for past block heights")
+        dump_output2 = n0.dumptxoutset('utxos2.dat', SNAPSHOT_BASE_HEIGHT)
+        check_dump_output(dump_output2)
```
added

fjahr commented:
Thanks for all the comments! Rebased and addressed feedback.

> One suggestion: on pruned nodes, the RPC should fail early if the block of the snapshot height is below the pruned block height

Added, good idea!

> I'm neutral about removing test_utxo_snapshots.sh. This seems to be more meant like an interactive demo to get familiar with AssumeUTXO and as such could still make sense for some?

Yeah, I was kind of aiming at that when I wrote "TBD if we need additional documentation.". Since your comment I have been thinking about it a bit more but because that script requires the person that runs it to change the code and recompile bitcoin core I think its audience was always just a tiny group, particularly I think it was useful for reviewers of the concept and code in the early days. Going forward it should be more valuable if we instead document the code properly to make it easy to grasp for contributors (the same goes for the functional tests where we essentially go through similar steps as the script) and for users we should provide easy documentation that explains the usage of the feature properly but doesn't require changing the code.

I have added another commit where I am splitting off a part of the previous design doc into a separate, usage-focused doc. Let me know what you think and if you have ideas for making the usage doc more complete.

> Seems like this should freeze indexes, wallets, etc, and hold a lock so nothing else happens during the process...?

> But rather than relying on block invalidation, I would prefer to have a clean rollback.

I think these requests are essentially the same, or at least the ideal solution to both of these will look similar.

For now, I have only added that I am setting the network to inactive and I will now work on a draft for the clean rollback. But I am still pushing this here now because I think this is already an improvement as-is and the clean rollback may be a bigger change that could also be done as a follow-up, so I don't think review needs to be blocked here. I will report back on the rollback progress here soon.

fjahr commented on src/rpc/blockchain.cpp:
```diff
@@ -2646,10 +2649,53 @@ static RPCHelpMan dumptxoutset()
+        if (node.chainman->m_blockman.IsPruneMode()) {
+            LOCK(::cs_main);
+            const CBlockIndex* current_tip{node.chainman->ActiveChain().Tip()};
+            const CBlockIndex* first_block{node.chainman->m_blockman.GetFirstStoredBlock(*current_tip)};
```
This actually isn't a sufficient check because we could be missing the undo data. I am addressing this here: #29668

fjahr replied:
This is fixed now since #29668 got merged and I am using this here now.

fjahr commented:
> Another approach, suggested by @maflcko in #29565 is to rewind the UTXO set inside of a cache. If that cache is kept isolated from the rest of the node, then indexes and wallets are not affected. It's also safe to abort the process, worst case just leaving a large temp file around.

Yes, I have a similar approach in mind. I think that is the cleanest solution if it's feasible in terms of complexity.

alfonsoromanz reviewed (approved):
Code Review and Tested ACK c83d7a5dec3595abe4cc6c8dcb29a016c6905a98. My impression is that this PR is good enough as is, unless there is a specific issue with `invalidate`/`reconsider` block that I am not aware of. 

This will not only prevent errors (by internalizing and testing code previously ran in scripts) but will also offer an easy way for users to create a snapshot from a specific height via RPC. If the rollback functionality gets implemented after this PR is merged, it seems straightforward to change this code to: (1) use the rollback instead of the `invalidateBlock`, and (2) remove the `ReconsiderBlock` call as the node will sync to the tip again. So I don't think the rollback feature should be a blocker for this PR.

Note: for testing I ran `test/functional/feature_assumeutxo.py` to make sure the test passes, and also modified line 391 to load the snapshot created by this new test in node2, to make sure the snapshot is loaded successfully too. i.e (line 391)
`loaded = n2.loadtxoutset(dump_output2['path'])`

fjahr commented:
Rebased

> My impression is that this PR is good enough as is, unless there is a specific issue with invalidate/reconsider block that I am not aware of.

> If the rollback functionality gets implemented after this PR is merged, it seems straightforward to change this code

Yes, either way, this PR should be an improvement. I am still planning to do the rollback in a cleaner way but if this PR is merged before that is ready, there should be no overhead resulting from that. It would become a refactoring follow-up that does not change the RPC interface again.

fjahr commented on src/rpc/blockchain.cpp:
```diff
@@ -2646,10 +2657,53 @@ static RPCHelpMan dumptxoutset()
+        InvalidateBlock(*node.chainman, invalidate_index->GetBlockHash());
+        const CBlockIndex* new_tip_index{WITH_LOCK(::cs_main, return node.chainman->ActiveChain().Tip())};
+
+        // In case there is any issue with a block being read from disc we need
```
Yes, interesting edge case! I extended the comment.

fjahr commented on src/rpc/blockchain.cpp:
```diff
@@ -2646,10 +2657,53 @@ static RPCHelpMan dumptxoutset()
+        // If the node is running in pruned mode we ensure all necessary block
+        // data is available before starting to roll back.
+        if (node.chainman->m_blockman.IsPruneMode()) {
+            LOCK(::cs_main);
```
Done

Sjors commented:
> I am still planning to do the rollback in a cleaner way but[...] It would become a refactoring follow-up that does not change the RPC interface again.

That makes sense.

ryanofsky commented on src/rpc/blockchain.cpp:
```diff
@@ -2615,6 +2615,9 @@ static RPCHelpMan dumptxoutset()
         "Write the serialized UTXO set to a file.",
```
In commit "RPC: Add height parameter to dumptxoutset" (https://github.com/bitcoin/bitcoin/commit/2cdce8fd9c93099db7e86ff33a1da8fbf0aaf364)

Documentation here is really terse. For a folluwup , I think it would be nice to add some of the documentation from the usage guide to the RPC documentation so it is more accessible.

fjahr replied:
Added a bit more detail in the docs improvement commit.

ryanofsky reviewed (approved):
Code review ACK 6a5408dd7da4c31283cb4f632b3a7aa692a53ca3. This is a really nice improvement over existing shell scripts, with a surprisingly simple implementation. I left some suggestions, but none are important and this looks good to merge in its current form.

fjahr commented on src/rpc/blockchain.cpp:
```diff
@@ -2654,10 +2657,56 @@ static RPCHelpMan dumptxoutset()
+
+        invalidate_index = WITH_LOCK(::cs_main, return node.chainman->ActiveChain().Next(target_index));
+        InvalidateBlock(*node.chainman, invalidate_index->GetBlockHash());
+        const CBlockIndex* new_tip_index{WITH_LOCK(::cs_main, return node.chainman->ActiveChain().Tip())};
```
Leaving this for a follow-up for now.

fjahr replied:
I have introduced your suggested fix as a separate commit because the main commit is already quite big and I think it's easier to review this way. I added some minor documentation changes since `CreateUTXOSnaphot()` is now only used in tests.

fjahr commented on src/rpc/blockchain.cpp:
```diff
@@ -2615,6 +2615,9 @@ static RPCHelpMan dumptxoutset()
         "Write the serialized UTXO set to a file.",
         {
             {"path", RPCArg::Type::STR, RPCArg::Optional::NO, "Path to the output file. If relative, will be prefixed by datadir."},
+            {"height", RPCArg::Type::NUM, RPCArg::DefaultHint{"the current tip"},
```
Right, I think that makes sense and this will be called mostly with this argument than anything else. That hadn't occurred to me so far honestly because I mostly wanted this for us to create snapshots that will potentially be added to the code in the future.

I have changed this to default to the latest valid snapshot height. That makes getting the tip more cumbersome for the users but I think that's ok since this should not be what most of the want, I think.

fjahr commented on doc/assumeutxo.md:
```diff
@@ -0,0 +1,65 @@
+from the genesis block. Once the background validation reaches the snapshot block,
+indexes will continue to build all the way to the tip.
+
+For indexes that support pruning, note that no pruning will take place between
+the snapshot and the tip, until the background sync has completed - after which
```
I need to spend some more time with this and test if this actually changed but this was added in #28618 which goes back to the conversation in the last assumeutxo PR: https://github.com/bitcoin/bitcoin/pull/27596#issuecomment-1545856135. I think nothing has changed in that regard though. I think the prune locks prevent the blocks of the snapshot from being pruned.

fjahr replied:
I made some edits here, making sure to clearly spell out what you suggested to include.

fjahr commented on doc/assumeutxo.md:
```diff
@@ -0,0 +1,65 @@
+
+There is currently no canonical source for snapshots, but any downloaded snapshot
+will be checked against a hash that's been hardcoded in source code. If there is
+no source for the snapshot you need, you can generate it yourself using
```
added and mentioned also that this node needs to be synced already

fjahr commented on src/rpc/blockchain.cpp:
```diff
@@ -2654,10 +2657,56 @@ static RPCHelpMan dumptxoutset()
+        InvalidateBlock(*node.chainman, invalidate_index->GetBlockHash());
+        const CBlockIndex* new_tip_index{WITH_LOCK(::cs_main, return node.chainman->ActiveChain().Tip())};
+
+        // In case there is any issue with a block being read from disc we need
```
fixed

fjahr commented on src/rpc/blockchain.cpp:
```diff
@@ -2654,10 +2657,56 @@ static RPCHelpMan dumptxoutset()
+
+        // Suspend network activity for the duration of the process when we are
+        // rolling back the chain to get a utxo set from a past height.
+        connman.SetNetworkActive(false);
```
I have implemented the RAII class and expanded the comment. If my understanding is correct I think we would punish a peer as misbehaving if they send us a block that is connecting to an invalid block (https://github.com/bitcoin/bitcoin/blob/master/src/net_processing.cpp#L1927) so we would punish any peer sending us a normal new block. We also wouldn't accept transactions that have a locktime above the snapshot height though I am not sure if switching off network activity is getting us a better result in that case.

fjahr commented:
Rebased and addressed feedback from @ryanofsky and @maflcko, thank you! This now includes the check for undo data since this was made possible by merge of https://github.com/bitcoin/bitcoin/pull/29668.

Asma1665 commented:
Revarse my assist

fjahr commented:
Fixed some outdated information in the remainin design doc and added that to the last commit.

ryanofsky reviewed (approved):
Code review ACK 27767184aaa7fa94600c70334cc4122e25872ff6.

Main change since last review is documenting dumptxoutset better and making default behavior be to dump the latest snapshot that can be loaded, instead of the current utxo set. Also improved other documentation and added NetworkDisable class to more reliably restore network state in case of error.

Left some additional suggestions but they aren't very important

mzumsande reviewed (commented):
Concept ACK - though I don't like having potentially large rollbacks as default behavior, see below.

ryanofsky commented on src/rpc/blockchain.cpp:
```diff
@@ -2695,10 +2717,66 @@ static RPCHelpMan dumptxoutset()
+    CConnman& connman = EnsureConnman(node);
+
+    const CBlockIndex* target_index{};
+    if (request.params[1].isNull()) {
```
> I don't like this default. This RPC is already kind of dangerous, it could roll back and forth the chain for hundred thousands of blocks (if executed a few years after it's released), which could take days

That's a good point. Given a choice of whether the default behavior might create a useless snapshot that can't be loaded or take the node offline trying to produce a really old snapshot, that makes me think there should probably not be a default value at all and the user should have to explicitly specify the behavior they want.

Would suggest adding a `type` parameter for the user to specify whether they want the latest snapshot or to roll back to a historic snapshot, and a `rollback` option to optionally specify the height or hash of a historic snapshot. Usage would look like:

```bash
bitcoin-cli dumptxoutset utxo.dat latest
bitcoin-cli dumptxoutset utxo.dat rollback
bitcoin-cli -named dumptxoutset utxo.dat rollback=853456
```

And a possible implementation would be:

<details><summary>diff</summary>
<p>

```diff
--- a/src/rpc/blockchain.cpp
+++ b/src/rpc/blockchain.cpp
@@ -2673,12 +2673,17 @@ static RPCHelpMan dumptxoutset()
     return RPCHelpMan{
         "dumptxoutset",
         "Write the serialized UTXO set to a file. This can be used in loadtxoutset afterwards if this snapshot height is supported in the chainparams as well.\n\n"
-        "Unless the requested height is the current tip, the node will roll back to the requested height and network activity will be suspended during this process."
+        "Unless the requested height is the current tip, the node will roll back to the requested height and network activity will be suspended during this process. "
         "Because of this it is discouraged to interact with the node in any other way during the execution of this call to avoid inconsistent results and race conditions, particularly RPCs that interact with blockstorage.",
         {
             {"path", RPCArg::Type::STR, RPCArg::Optional::NO, "Path to the output file. If relative, will be prefixed by datadir."},
-            {"height", RPCArg::Type::NUM, RPCArg::DefaultHint{"the latest valid snapshot height"},
-                "Height of the UTXO set file. Note: The further this number is from the tip, the longer this process will take. Consider setting a higher -rpcclienttimeout value in this case."
+            {"type", RPCArg::Type::STR, RPCArg::Default(""), "The type of snapshot to create. Can be \"latest\" to create a snapshot of the current UTXO set or \"rollback\" to temporarily roll back the state of the node to a historical block before creating the snapshot of a historical UTXO set. This parameter can be omitted if a separate \"rollback\" named parameter is specified indicating the height or hash of a specific historical block. If \"rollback\" is specified and separate \"rollback\" named parameter is not specified, this will roll back to the latest valid snapshot block that currently be loaded with loadtxoutset."},
+            {"options", RPCArg::Type::OBJ_NAMED_PARAMS, RPCArg::Optional::OMITTED, "",
+                {
+                    {"rollback", RPCArg::Type::NUM, RPCArg::Optional::OMITTED,
+                        "Height or hash of the block to roll back to before creating the snapshot. Note: The further this number is from the tip, the longer this process will take. Consider setting a higher -rpcclienttimeout value in this case.",
+                    RPCArgOptions{.skip_type_check = true, .type_str = {"", "string or numeric"}}},
+                },
             },
         },
         RPCResult{
@@ -2693,10 +2698,33 @@ static RPCHelpMan dumptxoutset()
                 }
         },
         RPCExamples{
-            HelpExampleCli("dumptxoutset", "utxo.dat")
+            HelpExampleCli("dumptxoutset", "utxo.dat latest") +
+            HelpExampleCli("dumptxoutset", "utxo.dat rollback") +
+            HelpExampleCli("-named dumptxoutset", R"(utxo.dat rollback=853456)")
         },
         [&](const RPCHelpMan& self, const JSONRPCRequest& request) -> UniValue
 {
+    NodeContext& node = EnsureAnyNodeContext(request.context);
+    const CBlockIndex* tip{WITH_LOCK(::cs_main, return node.chainman->ActiveChain().Tip())};
+    const CBlockIndex* target_index{nullptr};
+    const std::string snapshot_type{self.Arg<std::string>("type")};
+    const UniValue options{request.params[2].isNull() ? UniValue::VOBJ : request.params[2]};
+    if (options.exists("rollback")) {
+        if (!snapshot_type.empty() && snapshot_type != "rollback") {
+            throw JSONRPCError(RPC_INVALID_PARAMETER, strprintf("Invalid snapshot type \"%s\" specified with rollback option", snapshot_type));
+        }
+        target_index = ParseHashOrHeight(options["rollback"], *node.chainman);
+    } else if (snapshot_type == "rollback") {
+        auto snapshot_heights = node.chainman->GetParams().GetAvailableSnapshotHeights();
+        CHECK_NONFATAL(snapshot_heights.size() > 0);
+        auto max_height = std::max_element(snapshot_heights.begin(), snapshot_heights.end());
+        target_index = ParseHashOrHeight(*max_height, *node.chainman);
+    } else if (snapshot_type == "latest") {
+        target_index = tip;
+    } else {
+        throw JSONRPCError(RPC_INVALID_PARAMETER, strprintf("Invalid snapshot type \"%s\" specified. Please specify \"rollback\" or \"latest\"", snapshot_type));
+    }
+  
     const ArgsManager& args{EnsureAnyArgsman(request.context)};
     const fs::path path = fsbridge::AbsPathJoin(args.GetDataDirNet(), fs::u8path(request.params[0].get_str()));
     // Write to a temporary path and then move into `path` on completion
@@ -2718,19 +2746,7 @@ static RPCHelpMan dumptxoutset()
             "Couldn't open file " + temppath.utf8string() + " for writing.");
     }
 
-    NodeContext& node = EnsureAnyNodeContext(request.context);
     CConnman& connman = EnsureConnman(node);
-
-    const CBlockIndex* target_index{};
-    if (request.params[1].isNull()) {
-        auto snapshot_heights = node.chainman->GetParams().GetAvailableSnapshotHeights();
-        auto max_height = std::max_element(snapshot_heights.begin(), snapshot_heights.end());
-        target_index = ParseHashOrHeight(*max_height, *node.chainman);
-    } else {
-        target_index = ParseHashOrHeight(request.params[1], *node.chainman);
-    }
-
-    const auto tip{WITH_LOCK(::cs_main, return node.chainman->ActiveChain().Tip())};
     const CBlockIndex* invalidate_index{nullptr};
     std::unique_ptr<NetworkDisable> disable_network;
 
--- a/src/rpc/client.cpp
+++ b/src/rpc/client.cpp
@@ -185,7 +185,8 @@ static const CRPCConvertParam vRPCConvertParams[] =
     { "gettxoutproof", 0, "txids" },
     { "gettxoutsetinfo", 1, "hash_or_height" },
     { "gettxoutsetinfo", 2, "use_index"},
-    { "dumptxoutset", 1, "height"},
+    { "dumptxoutset", 2, "options" },
+    { "dumptxoutset", 2, "rollback" },
     { "lockunspent", 0, "unlock" },
     { "lockunspent", 1, "transactions" },
     { "lockunspent", 2, "persistent" },
```

</p>
</details>

fjahr replied:
Thanks, I agree that the default wasn't very sensitive when looking at it this way. I have adopted the type approach that @ryanofsky suggested. I think it may be a bit more complicated to understand initially for users but it's also easier to hand because you don't have to deal with a hash or height in the most common use cases.

fjahr committed:
rpc, refactor: Prevent potential race conditions in dumptxoutset

Co-authored-by: Ryan Ofsky <ryan@ofsky.org>

fjahr commented on src/rpc/blockchain.cpp:
```diff
@@ -2672,7 +2672,9 @@ static RPCHelpMan dumptxoutset()
         "dumptxoutset",
-        "Write the serialized UTXO set to a file.",
+        "Write the serialized UTXO set to a file. This can be used in loadtxoutset afterwards if this snapshot height is supported in the chainparams as well.\n\n"
+        "Unless the requested height is the current tip, the node will roll back to the requested height and network activity will be suspended during this process."
```
fixed

fjahr commented on src/rpc/blockchain.cpp:
```diff
@@ -2687,10 +2719,68 @@ static RPCHelpMan dumptxoutset()
+    if (request.params[1].isNull()) {
+        auto snapshot_heights = node.chainman->GetParams().GetAvailableSnapshotHeights();
+        auto max_height = std::max_element(snapshot_heights.begin(), snapshot_heights.end());
+        target_index = ParseHashOrHeight(*max_height, *node.chainman);
```
fixed (as part of one of the larger changes)

fjahr commented on src/rpc/blockchain.cpp:
```diff
@@ -2656,6 +2675,9 @@ static RPCHelpMan dumptxoutset()
             {"path", RPCArg::Type::STR, RPCArg::Optional::NO, "Path to the output file. If relative, will be prefixed by datadir."},
+            {"height", RPCArg::Type::NUM, RPCArg::DefaultHint{"the latest valid snapshot height"},
+                "Height of the UTXO set file. Note: The further this number is from the tip, the longer this process will take. Consider setting a higher -rpcclienttimeout value in this case."
+            },
```
Adopted the type approach you suggested so this is resolved as well with that.

fjahr commented:
Addressed feedback and rebased.

There is a bit of a hack in the tests now that is made necessary by an unrelated bug: #26245 I marked this with an TODO accordingly.

ryanofsky commented on src/rpc/blockchain.cpp:
```diff
@@ -2774,44 +2790,52 @@ static RPCHelpMan dumptxoutset()
-UniValue CreateUTXOSnapshot(
-    NodeContext& node,
+std::tuple<std::unique_ptr<CCoinsViewCursor>, CCoinsStats, const CBlockIndex*>
+PrepareUTXOSnapshot(
```
In commit "rpc, refactor: Prevent potential race conditions in dumptxoutset" (0c1238917a5edf6bdcae15aab5f91b1bdd2540b9)

Would be good to update comment on line 2852. Could replace "use below this block" with "use in WriteUTXOSnapshot"

fjahr replied:
Done

ryanofsky commented on src/rpc/blockchain.cpp:
```diff
@@ -2774,44 +2790,52 @@ static RPCHelpMan dumptxoutset()
+        result = WriteUTXOSnapshot(*chainstate, cursor.get(), &stats, tip, afile, path, temppath, node.rpc_interruption_point);
+        fs::rename(temppath, path);
+    }
     if (invalidate_index) {
```
In commit "rpc, refactor: Prevent potential race conditions in dumptxoutset" (0c1238917a5edf6bdcae15aab5f91b1bdd2540b9)

Would it make any sense to call `ReconsiderBlock()` and `disable_network.reset()` before calling `WriteUTXOSnapshot` instead of after?

Not sure if that might make the node more usable while the snapshot is being created, or just slow down writing the snapshot and increase resource usage.

Change wouldn't be appropriate in this commit anyway, but could be a possible followup or possible future `dumptxoutset` RPC option if it makes sense.

fjahr replied:
I did some light testing of this and didn't see a negative performance impact. Actually the three times I ran with this change were slightly faster on average than with the current code here. But there was also quite a bit of volatility so that might have been pure luck. It doesn't look like it slows things down though, at least not on my machine.

I have left it out for now but can add it. I am curious what you mean by having it as an RPC option. This seems like something that we should just do if it's safe and we think it's beneficial for users and if not we just leave it out. I'm not sure which users would care enough about such a setting. This is also not a RPC that a lot of people will use when they do it's rather frequently I assume.

Sjors replied:
It seems conceptually simpler to first write the snapshot, reconsider all blocks that were marked invalid and only then restore network activity.

>  make the node more usable while the snapshot is being created

Lower usability is probably feature. The user should not be doing anything important until the chain is back at the original tip, e.g. loading a wallet could still cause confusion.

ryanofsky replied:
re: https://github.com/bitcoin/bitcoin/pull/29553#discussion_r1725892812

Interesting results and I agree it makes sense to leave this alone for now. I thought it adding an option might be reasonable because doing this might increase resource usage even if it is faster. And maybe it is better if node is unusable for longer as Sjors says. I don't have a clear sense of what is best here.

ryanofsky reviewed (approved):
Code review ACK 0c1238917a5edf6bdcae15aab5f91b1bdd2540b9. Main change since last review is that dumptxoutset will no longer dump a snapshot that can't be loaded by default or roll back the chain by default. Instead you need to specify "latest" or "rollback" to choose the behavior you want, or specify the height or hash of a specific block.

It is now possible to specify a snapshot block hash instead of a just height, and there are test and documentation improvements, and a new commit that locks cs_main between rolling back the chainstate and getting the UTXO database cursor to guarantee the right snapshot block is always used.

pablomartin4btc reviewed (commented):
Concept ACK

I agree with the reasoning behind removing the scripts, specially after founding this [bug](https://github.com/bitcoin/bitcoin/pull/28553#pullrequestreview-2251032570) during testing. I'll also test this PR soon.

fjahr commented on src/rpc/blockchain.cpp:
```diff
@@ -2774,44 +2790,52 @@ static RPCHelpMan dumptxoutset()
+    UniValue result;
+    UniValue error;
+    {
+        LOCK(node.chainman->GetMutex());
```
Taken, thanks!

fjahr commented:
Improved comments based on suggestions by @ryanofsky , thanks a lot!

Sjors commented on src/rpc/blockchain.cpp:
```diff
@@ -2697,11 +2747,55 @@ static RPCHelpMan dumptxoutset()
+        // to stop here, otherwise the dump could still be created for the wrong
+        // height.
+        // The new tip could also not be the target block if we have a stale
+        // sister block of invalidate_index. This block (or a descendant) would
```
b23abfca68f7b349e6d2f9e04369f0fdc82d2a98: a followup could handle multiple blocks at the same height by running a loop that invalidates all (new) tips at the target height that are not the block hash we expected.

Those could be tracked in `invalidate_indexes` and be reconsidered in reverse order (and with the network still off) after the snapshot is made.

fjahr replied:
Leaving this for a follow-up.

Sjors replied:
It can wait until some bully actually wastes $100K+ to produce such a stale block at an existing assume utxo height :-)

Sjors commented on src/rpc/blockchain.cpp:
```diff
@@ -2697,11 +2747,55 @@ static RPCHelpMan dumptxoutset()
+        // sister block of invalidate_index. This block (or a descendant) would
+        // be activated as the new tip and we would not get to new_tip_index.
+        if (new_tip_index != target_index) {
+            ReconsiderBlock(*node.chainman, invalidate_index->GetBlockHash());
```
b23abfca68f7b349e6d2f9e04369f0fdc82d2a98: assuming this is blocking and it could take hours before the throw below, it would be good to emit a log message in the mean time.

fjahr replied:
done

Sjors commented on src/rpc/blockchain.cpp:
```diff
@@ -2697,11 +2747,55 @@ static RPCHelpMan dumptxoutset()
+    }
+
     UniValue result = CreateUTXOSnapshot(
         node, node.chainman->ActiveChainstate(), afile, path, temppath);
```
b23abfca68f7b349e6d2f9e04369f0fdc82d2a98: for a followup, please check if this path is writable early in the RPC function. I once wasted a few hours because I used `~/someplace` which it didn't understand.

fjahr replied:
Hm, doesn't this code catch this already?

```
    FILE* file{fsbridge::fopen(temppath, "wb")};
    AutoFile afile{file};
    if (afile.IsNull()) {
```

I get this here:

```
$ src/bitcoin-cli dumptxoutset \~/Downloads/utxo.dat latest
error code: -8
error message:
Couldn't open file /Users/XXX/Library/Application Support/Bitcoin/~/Downloads/utxo.dat.incomplete for writing.
```

Can you give a more detailed example to reproduce this maybe?

Sjors replied:
I ran into that issue with the bash script, not this PR. This PR actually fixes the problem. The bash script would first do the rollback and only then call `dumptxoutset`, which is where the check is. So it's fine now.

Sjors reviewed (approved):
tACK b5292334e5e8792644123b627896359eb1da8d25

I was initially confused because I got this:

```
$ src/bitcoin-cli dumptxoutset $HOME/dev/utxo-snapshots/utxo-840000-pr29553.dat rollback
error code: -1
error message:
Internal bug detected: snapshot_heights.size() > 0
rpc/blockchain.cpp:2738 (operator())
Bitcoin Core v27.99.0-b5292334e5e8
Please report this issue here: https://github.com/bitcoin/bitcoin/issues
````

But that's because this PR wasn't rebased after #28553 got merged.

After rebasing it myself, I was able to reproduce the snapshot at height 840,000.

This can wait for a followup, but can you add `-rpcclienttimeout=0` to all the examples? And to the help text:

```
"This call may take several minutes. Make sure to use no RPC timeout (bitcoin-cli -rpcclienttimeout=0)",
```

The one thing I like about the original script, which the RPC can't give us, is a progress report.

This is a general problem with long running RPC calls.

I'm a bit confused about the last commit, especially @ryanofsky's comment:

```cpp
// The cursor will remain valid and be used by
// WriteUTXOSnapshot to write a consistent snapshot even if the
// chainstate changes.
```

Does `CCoinsViewCursor` (and LevelDB) somehow guarantee that if coins are deleted from the chainstate database, as new blocks are processed, we can still access them? (and that we won't see newly added coins).

Or merely that the cursor becomes invalid so we can detect that things changed from under us? If so, where is that handled?

For the (main) use case of rolling back to a specific height it doesn't matter, because we first create and save the snapshot and only then call ReconsiderBlock. But for the case of making a snapshot at the tip, it's not clear to me what would happen.

ryanofsky reviewed (approved):
Code review ACK b5292334e5e8792644123b627896359eb1da8d25, just updated code comments since last review.

---

re: https://github.com/bitcoin/bitcoin/pull/29553#pullrequestreview-2253922313

> Does CCoinsViewCursor (and LevelDB) somehow guarantee that if coins are deleted from the chainstate database, as new blocks are processed, we can still access them? (and that we won't see newly added coins).

CCoinsViewCursor is a dumb cursor, so leveldb would be doing all the work, but my understanding is leveldb cursors iterate over the database at a snapshot in time and don't reflect future updates or deletions.

fjahr commented on src/rpc/blockchain.cpp:
```diff
@@ -2659,6 +2678,14 @@ static RPCHelpMan dumptxoutset()
         "Write the serialized UTXO set to a file.",
         {
             {"path", RPCArg::Type::STR, RPCArg::Optional::NO, "Path to the output file. If relative, will be prefixed by datadir."},
+            {"type", RPCArg::Type::STR, RPCArg::Default(""), "The type of snapshot to create. Can be \"latest\" to create a snapshot of the current UTXO set or \"rollback\" to temporarily roll back the state of the node to a historical block before creating the snapshot of a historical UTXO set. This parameter can be omitted if a separate \"rollback\" named parameter is specified indicating the h ...
```
I agree that `-height=XXXX` is better but having `-rollback` and `-height=XXXXX` seems worse than `rollback` and `-rollback=XXXXX` and of course just `-height` doesn't make sense. So leaving this as is for now.

fjahr commented on src/rpc/blockchain.cpp:
```diff
@@ -2675,7 +2675,9 @@ static RPCHelpMan dumptxoutset()
         "dumptxoutset",
-        "Write the serialized UTXO set to a file.",
+        "Write the serialized UTXO set to a file. This can be used in loadtxoutset afterwards if this snapshot height is supported in the chainparams as well.\n\n"
+        "Unless the the latest type is requested, the node will roll back to the requested height and network activity will be suspended during this process. "
```
done

fjahr commented on doc/assumeutxo.md:
```diff
@@ -0,0 +1,79 @@
+state that does not actually reflect reality, i.e. blocks are marked valid
+although we know they are not invalid. Because of this it is discouraged to
+interact with the node in any other way during this time to avoid inconsistent
+results and race conditions, particularly RPCs that interact with blockstorage.
```
added

fjahr commented on doc/assumeutxo.md:
````diff
@@ -0,0 +1,79 @@
+```
+
+For most of the duration of `dumptxoutset` running the node is in a temporary
+state that does not actually reflect reality, i.e. blocks are marked valid
````
fixed

fjahr commented:
I addressed @Sjors 's feedback and rebased so the behavior can be tested with the mainnet params.

> This can wait for a followup, but can you add -rpcclienttimeout=0 to all the examples?

Done

> But that's because this PR wasn't rebased after https://github.com/bitcoin/bitcoin/pull/28553 got merged.

Rebased for easier testing.

Sjors commented:
re-utACK ac9d53d0dee26db58ab125aa369f476c232da660

ryanofsky reviewed (approved):
Code review ACK ac9d53d0dee26db58ab125aa369f476c232da660. Nice documentation improvements since last review, and a new log statement

fjahr commented:
Rebased due to #30690

Sjors commented:
re-utACK 46c8d75d24d3355ec468f7f608effe94636e16db

#30690 just modified the script this PR deletes.

pablomartin4btc reviewed (approved):
tACK 46c8d75d24d3355ec468f7f608effe94636e16db

I've reviewed the code, it looks very neat.

<details>
<summary>I've tested <code>dumptxoutset</code> with both argument types, "<code>latest</code>" (network is still active on this "mode") and "<code>rollback</code>" (network gets temporarily disabled), plus specifiyng .</summary>

```
/src/bitcoin-cli -datadir=${AU_DATADIR} -rpcclienttimeout=0 dumptxoutset utxo-840000.dat rollback
{
  "coins_written": 176948713,
  "base_hash": "0000000000000000000320283a032748cef8227873ff4872689bf23f1cda83a5",
  "base_height": 840000,
  "path": "/home/pablo/.test_utxo_840/utxo-840000.dat",
  "txoutset_hash": "a2a5521b1b5ab65f67818e5e8eccabb7171a517f9e2382208f77687310768f96",
  "nchaintx": 991032194
}


./src/bitcoin-cli -datadir=${AU_DATADIR} -rpcclienttimeout=0 dumptxoutset utxo-859000.dat latest
{
  "coins_written": 185421121,
  "base_hash": "00000000000000000002806cbe89e9694072f8f2a04cc77b73f69a4c9cdc5ec3",
  "base_height": 859144,
  "path": "/home/pablo/.test_utxo_840/utxo-859000.dat",
  "txoutset_hash": "2c0d181bcbef8364ac4202671960e91484ee3b64d33a72df35f4ffe5064852ae",
  "nchaintx": 1070242632
}

./src/bitcoin-cli -datadir=${AU_DATADIR} -rpcclienttimeout=0 -named dumptxoutset utxo-858000.dat rollback=858000
{
  "coins_written": 184931392,
  "base_hash": "00000000000000000001b44004f45de683f7f5dc792519b47f867baac2d4bf56",
  "base_height": 858000,
  "path": "/home/pablo/.test_utxo_840/utxo-858000.dat",
  "txoutset_hash": "d0b6dd9367746570658e9052ff814b864643d8aeccb2ccc81d438123242db63b",
  "nchaintx": 1064936785
}

```


</details>

As @Sjors mentioned above please consider adding -rpcclienttimeout=0 to the calls also in the documentation. Furthermore, these example outputs could be useful there too.

fjahr commented:
> > As @Sjors mentioned above please consider adding -rpcclienttimeout=0 to the calls also in the documentation. Furthermore, these example outputs could be useful there too.
> 
> Can you say which examples you mean? I added it to the RPC help examples in my last push but I guess you might be referring to the README examples where I didn't do it yeah. Or is there another place I am still missing?

I just added the timeout to the README example as well and rebased since that was recommend for all PRs anyway in the light of the CMake introduction.

Should be trivial to re-ack with

```
$ git range-diff master 46c8d75d24d3355ec468f7f608effe94636e16db 94b0adcc371540732453d70309c4083d4bd9cd6b
```

Sjors commented:
re-utACK 94b0adcc371540732453d70309c4083d4bd9cd6b

Just a rebase and documentation improvement.

mzumsande reviewed (commented):
ACK 94b0adcc371540732453d70309c4083d4bd9cd6b

It would be good to fix the `networkactive` issue below, here or in a follow-up.

pablomartin4btc commented on src/rpc/blockchain.cpp:
```diff
@@ -2698,11 +2748,55 @@ static RPCHelpMan dumptxoutset()
+        // this so we don't punish peers that send us that send us data that
+        // seems wrong in this temporary state. For example a normal new block
+        // would be classified as a block connecting an invalid block.
+        disable_network = std::make_unique<NetworkDisable>(connman);
```
I do agree unless there's another use case that invalidates it.

MarcusMWilliams commented:
1-225-316-0850

On Mon, Sep 2, 2024, 12:47 PM pablomartin4btc ***@***.***>
wrote:

> ***@***.**** commented on this pull request.
> ------------------------------
>
> In src/rpc/blockchain.cpp
> <https://github.com/bitcoin/bitcoin/pull/29553#discussion_r1741151732>:
>
> > +        // data is available before starting to roll back.
> +        if (node.chainman->m_blockman.IsPruneMode()) {
> +            LOCK(node.chainman->GetMutex());
> +            const CBlockIndex* current_tip{node.chainman->ActiveChain().Tip()};
> +            const CBlockIndex* first_block{node.chainman->m_blockman.GetFirstBlock(*current_tip, /*status_mask=*/BLOCK_HAVE_MASK)};
> +            if (first_block->nHeight > target_index->nHeight) {
> +                throw JSONRPCError(RPC_MISC_ERROR, "Could not roll back to requested height since necessary block data is already pruned.");
> +            }
> +        }
> +
> +        // Suspend network activity for the duration of the process when we are
> +        // rolling back the chain to get a utxo set from a past height. We do
> +        // this so we don't punish peers that send us that send us data that
> +        // seems wrong in this temporary state. For example a normal new block
> +        // would be classified as a block connecting an invalid block.
> +        disable_network = std::make_unique<NetworkDisable>(connman);
>
> I do agree unless there's another use case that invalidates it.
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/bitcoin/bitcoin/pull/29553#discussion_r1741151732>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ASHJW54AUQI54MR5QF3EEK3ZUSQBLAVCNFSM6AAAAABEFOZY32VHI2DSMVQWIX3LMV43YUDVNRWFEZLROVSXG5CSMV3GSZLXHMZDENZWGAZDEMRYGM>
> .
> You are receiving this because you are subscribed to this thread.Message
> ID: ***@***.***>
>

pablomartin4btc commented:
> > As @Sjors mentioned above please consider adding -rpcclienttimeout=0 to the calls also in the documentation. Furthermore, these example outputs could be useful there too.
> 
> Can you say which examples you mean? I added it to the RPC help examples in my last push but I guess you might be referring to the README examples where I didn't do it yet. Or is there another place I am still missing?

Thanks! I've checked the places where you added the `-rpcclienttimeout=0`, also it could be added into the `loadtxoutset` rpc call in the same file (`/doc/assumeutxo.md`), as it was suggested by @Sjors in the [issue](https://github.com/bitcoin/bitcoin/issues/28620#issue-1933055748) he opened to make the rpc async at some point. 

The other thing I was suggesting was to add the outputs (or a part of it) of the commands I ran in my review, just to the user to see what to expect and compare the hash, etc. But it's up to you ofc.

pablomartin4btc reviewed (approved):
> `$ git range-diff master 46c8d75d24d3355ec468f7f608effe94636e16db 94b0adcc371540732453d70309c4083d4bd9cd6b
`

re-ACK 94b0adcc371540732453d70309c4083d4bd9cd6b

fjahr commented:
There are now 3 ACKs so unless need to retouch I will make a quick follow-up addressing the open comments from @pablomartin4btc and @mzumsande .

achow101 commented:
ACK 46c8d75d24d3355ec468f7f608effe94636e16db

fjahr commented:
Follow-ups are here: https://github.com/bitcoin/bitcoin/pull/30808
